### PR TITLE
Add named type wildcards

### PR DESCRIPTION
**Pull request by natefaubion**
Implements #715 #1228

This is a slight UX breaking change. This adds `?foo` syntax for types, which act as named wildcards which always warn. The previous `_` syntax acts as a partial type signature, which does not warn unless it's part of a module-level type signature.

I didn't change the warning (`WildcardInferredType`) except to add the name, but maybe it warrants a separate warning.

### CODE REVIEW

File: src/Language/PureScript/TypeChecker.hs [hunk 1]
```diff
@@ -43,6 +43,8 @@ import Language.PureScript.Types
 
 import Lens.Micro.Platform ((^..), _1, _2)
 
+import Debug.Trace as T
```
Comment (LiamGoodacre) [inline src/Language/PureScript/TypeChecker.hs]: (Debugging artefact)

**Comment by natefaubion** [comment]
One UX consideration is should `?foo` in types emit an error rather than a warning? It's necessary for values, but not for types. If we leave it as a warning, that means you can accidentally leave `?foo` in a module inadvertently since we don't persist warnings.

**Comment by LiamGoodacre** [comment]
I'd prefer if both type-level and term-level holes behaved the same, so if term holes error then so should type holes.

**Comment by LiamGoodacre** [comment]
To add to that - I feel like `_` shouldn't ever error or warn, and `?foo` should always error.
Now that we have the option to use `?foo`, the intent behind using `_` becomes "fill in this bit for me" instead of "help me figure this bit out".

**Comment by natefaubion** [comment]
Currently we warn if you leave off a type signature for module-level declarations, which is why I preserved the `_` warning there. Otherwise, `foo = 42 :: _` just becomes a way to not get warnings while still leaving off signature.

**Comment by rightfold** [comment]
> I'd prefer if both type-level and term-level holes behaved the same, so if term holes error then so should type holes.

And they should be called _kinded holes_ and report only the kind. :sweat_smile:

**Comment by natefaubion** [comment]
This is ready for review again. I've addressed the UX concerns.

* `?foo` always raises a `HoleInferredType` error. When it's in a type signature, it just does not provide a `TypeSearch`.
* `_` does not warn, unless it's in a module-level type signature, where it continues to throw a `WildcardInferredType` warning.

**Pull request merged**